lib/AST/Decl.cpp

```diff
@@ -5033,10 +5033,40 @@ findGenericParameterReferencesRec(CanGenericSignature genericSig,
     return info;
   }
 
-  if (!type->isTypeParameter()) {
+  // Packs are invariant.
+  if (auto *pack = type->getAs<PackType>()) {
+    auto info = GenericParameterReferenceInfo();
+
+    for (auto arg : pack->getElementTypes()) {
+      info |= findGenericParameterReferencesRec(
+          genericSig, genericParam, arg,
+          TypePosition::Invariant, /*canBeCovariantResult=*/false);
+    }
+
+    return info;
+  }
+
+  // Pack expansions are invariant.
+  if (auto *expansion = type->getAs<PackExpansionType>()) {
+    return findGenericParameterReferencesRec(
+        genericSig, genericParam, expansion->getPatternType(),
+        TypePosition::Invariant, /*canBeCovariantResult=*/false);
+  }
+
+  // Specifically ignore parameterized protocols and existential
+  // metatypes because we can erase them to the upper bound.
+  if (type->is<ParameterizedProtocolType>() ||
+      type->is<ExistentialMetatypeType>()) {
```

Comment:
@slavapestov I think we need to be more careful with this sort of loosening. The more sophisticated erasure we support, the harder it will be for everyone to get off of it once we come up with something better. The non-metatype case was never supported, so I think the same should hold for an existential metatype:

```swift
class C<T> {}
protocol P {
  associatedtype A

  func f() -> any P & C<A>
  func fMeta() -> any (P & C<A>).Type
}

do {
  let p: any P
  let _ = p.f() // error
  let _ = p.fMeta() // crash
}
```

The call to `fMeta()` used to miscompile and now crashes, so we can still make it an error.

Reply:
#78593